sbncode/BeamSpillInfoRetriever/NuMIRetriever/NuMIRetriever_module.cc

```diff
@@ -0,0 +1,254 @@
+
+#include "IFBeam_service.h"
+#include "ifbeam_c.h"
+//#include "MWRData.h"
+
+#include <memory>
+#include <optional>
+#include <vector>
+
+namespace sbn {
+  class NuMIRetriever;
+}
+
+class sbn::NuMIRetriever : public art::EDProducer {
+public:
+  explicit NuMIRetriever(fhicl::ParameterSet const &p);
+  NuMIRetriever(NuMIRetriever const&) = delete;
+  NuMIRetriever(NuMIRetriever &&) = delete;
+  NuMIRetriever& operator=(NuMIRetriever const&) = delete;
+  NuMIRetriever& operator=(NuMIRetriever&&) = delete;
+  
+  // Required functions.
+  void produce(art::Event& e) override;
+  void beginSubRun(art::SubRun& sr);
+  void endSubRun(art::SubRun& sr);
+  
+private:
+  // input labels
+  std::vector< sbn::NuMISpillInfo > fOutbeamInfos;
+  double fTimePad;
+  double fBFPEpsilion;
+  std::string fURL;
+  //MWRData mwrdata;
+  std::string raw_data_label_;
+  std::string fDeviceUsedForTiming;
+  int TotalBeamSpills;
+  art::ServiceHandle<ifbeam_ns::IFBeam> ifbeam_handle;
+  std::unique_ptr<ifbeam_ns::BeamFolder> bfp;
+};
+
+sbn::NuMIRetriever::NuMIRetriever(fhicl::ParameterSet const& p)
+  : EDProducer{p},
+  fTimePad(p.get<double>("TimePadding", 0.5)), //epsilon in seconds, buffer time to look for spills
+  fBFPEpsilion(p.get<double>("BFPEpsilon", 0.02)), // 20 ms, tuned for BNB, check for NuMI here might need to be larger
+  raw_data_label_(p.get<std::string>("raw_data_label")),
+  fDeviceUsedForTiming(p.get<std::string>("DeviceUsedForTiming")),
+  bfp(ifbeam_handle->getBeamFolder(p.get<std::string>("Bundle"), p.get<std::string>("URL"), p.get<double>("TimeWindow")))
+{
+
+  bfp->set_epsilon(fBFPEpsilion);
+  produces<std::vector<sbn::NuMISpillInfo>, art::InSubRun>();
+  TotalBeamSpills = 0;
+}
+
+void sbn::NuMIRetriever::produce(art::Event &e)
+{
+
+  // If this is the first event in the run, then ignore it
+  // We do not currently have the ability to figure out the first
+  // spill that the DAQ was sensitive to, so don't try to save any
+  // spill information
+  //
+  // TODO: long-term goal -- can we fix this?
+  if (e.event() == 1) return;
+
+  int gate_type = 0;
+  art::Handle< std::vector<artdaq::Fragment> > raw_data_ptr;
+  e.getByLabel(raw_data_label_, "ICARUSTriggerUDP", raw_data_ptr);
+  auto const & raw_data = (*raw_data_ptr);
+
+  double t_current_event  = 0;
+  double t_previous_event = 0;
+  double number_of_gates_since_previous_event = 0;
+
+  for(auto raw_datum : raw_data){
+
+    uint64_t artdaq_ts = raw_datum.timestamp();
+    icarus::ICARUSTriggerUDPFragment frag(raw_datum);
+    std::string data = frag.GetDataString();
+    char *buffer = const_cast<char*>(data.c_str());
+    icarus::ICARUSTriggerInfo datastream_info = icarus::parse_ICARUSTriggerString(buffer);
+    gate_type = datastream_info.gate_type;
+    number_of_gates_since_previous_event = frag.getDeltaGatesNuMI();
+
+    t_current_event = static_cast<double>(artdaq_ts)/(1000000000); //check this offset... 
+    if(gate_type == 2)
```

Comment:
I am not sure if this is the right gate type, can we find someone to confirm

Reply:
This is still the case

Reply:
@jzettle?

Reply:
Also these files are split by the gate type so if this is incorrect you should have seen zero spills if I recall correctly